src/Module.ts

```diff
@@ -716,6 +716,27 @@ export default class Module {
 		this.includeAllExports(false);
 	}
 
+	includeExportsByNames(names: readonly string[], includeNamespaceMembers: boolean): void {
+		if (!this.isExecuted) {
+			markModuleAndImpureDependenciesAsExecuted(this);
+			this.graph.needsTreeshakingPass = true;
+		}
+
+		for (const name of names) {
+			const variable = this.getVariableForExportName(name)[0];
+			if (variable) {
+				variable.deoptimizePath(UNKNOWN_PATH);
+				if (!variable.included) {
+					this.includeVariable(variable);
+				}
+			}
+		}
+
+		if (includeNamespaceMembers) {
```

Comment:
Setting this always to `true` is fine, but there are cases where we can actually set this to `false`. Let me explain.
There are two situations where the namespace will contain exports that we do not know at build time:

1. External namespace reexports `export * from 'external'`
2. Synthetic namespaces, i.e. unresolved exports will be picked as properties from an object export.

In both cases, those exports have lower priority than any existing named exports. So if we only import variables that are actually exported in the module, or reexported from another module by name. To check for this, you can verify

```js
this.exports.has(name) || this.reexportDescriptions.has(name);
```

If that is the case for all names, you can skip the "merged namespaces". So why would this be beneficial, and how can you test this? This is relevant in situations where the namespace is rendered as an object. For dynamic imports, this is always the case if the file has additional exports that are not part of the namespace. Then it gets complicated, see this test-case:
https://rollup-ebmohmssk-rollup-js.vercel.app/repl/?pr=4952&shareable=JTdCJTIyZXhhbXBsZSUyMiUzQW51bGwlMkMlMjJtb2R1bGVzJTIyJTNBJTVCJTdCJTIyY29kZSUyMiUzQSUyMmNvbnN0JTIwJTdCZm9vJTJDeCU3RCUyMCUzRCUyMGF3YWl0JTIwaW1wb3J0KCcuJTJGZGVwMS5qcycpJTVDbmNvbnNvbGUubG9nKGZvbyklM0IlMjIlMkMlMjJpc0VudHJ5JTIyJTNBdHJ1ZSUyQyUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTdEJTJDJTdCJTIyY29kZSUyMiUzQSUyMmV4cG9ydCUyMGNvbnN0JTIwZm9vJTIwJTNEJTIwMyUzQiU1Q25leHBvcnQlMjBjb25zdCUyMGJhciUyMCUzRCUyMDQlM0IlNUNuZXhwb3J0JTIwKiUyMGZyb20lMjAnZXh0ZXJuYWwnJTVDbiU1Q25pbXBvcnQoJy4lMkZkZXAyJyklM0IlNUNuaW1wb3J0JTIwJTdCc2hhcmVkJTdEJTIwZnJvbSUyMCcuJTJGc2hhcmVkJyUzQiU1Q25jb25zb2xlLmxvZyhzaGFyZWQpJTNCJTIyJTJDJTIyaXNFbnRyeSUyMiUzQWZhbHNlJTJDJTIybmFtZSUyMiUzQSUyMmRlcDEuanMlMjIlN0QlMkMlN0IlMjJjb2RlJTIyJTNBJTIyaW1wb3J0JTIwJy4lMkZkZXAxJyU1Q25pbXBvcnQlMjAlN0JzaGFyZWQlN0QlMjBmcm9tJTIwJy4lMkZzaGFyZWQnJTNCJTVDbmNvbnNvbGUubG9nKHNoYXJlZCklM0IlMjIlMkMlMjJpc0VudHJ5JTIyJTNBZmFsc2UlMkMlMjJuYW1lJTIyJTNBJTIyZGVwMi5qcyUyMiU3RCUyQyU3QiUyMmNvZGUlMjIlM0ElMjJleHBvcnQlMjBjb25zdCUyMHNoYXJlZCUyMCUzRCUyMCdmb28nJTNCJTIyJTJDJTIyaXNFbnRyeSUyMiUzQWZhbHNlJTJDJTIybmFtZSUyMiUzQSUyMnNoYXJlZC5qcyUyMiU3RCU1RCUyQyUyMm9wdGlvbnMlMjIlM0ElN0IlN0QlN0Q=

Here the dynamic import target has itself a dynamic import, and both imported files share a dependency. That dependency is hoisted into the first import target, tainting the namespace. In the example, we restructure `x` that does not exist. If we would not do that, we would not need any of the mergedNamespace helper logic. To see what this would look like, just remove the `export * from 'external'` logic.

Not sure if you want to add this optimization, but at least it is not costly.

Reply:
Ah, I just noticed a much simpler test case is to use `inlineDynamicImports`: https://rollup-ebmohmssk-rollup-js.vercel.app/repl/?pr=4952&shareable=JTdCJTIyZXhhbXBsZSUyMiUzQW51bGwlMkMlMjJtb2R1bGVzJTIyJTNBJTVCJTdCJTIyY29kZSUyMiUzQSUyMmNvbnN0JTIwJTdCZm9vJTJDeCU3RCUyMCUzRCUyMGF3YWl0JTIwaW1wb3J0KCcuJTJGZGVwMS5qcycpJTVDbmNvbnNvbGUubG9nKGZvbyklM0IlMjIlMkMlMjJpc0VudHJ5JTIyJTNBdHJ1ZSUyQyUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTdEJTJDJTdCJTIyY29kZSUyMiUzQSUyMmV4cG9ydCUyMGNvbnN0JTIwZm9vJTIwJTNEJTIwMyUzQiU1Q25leHBvcnQlMjBjb25zdCUyMGJhciUyMCUzRCUyMDQlM0IlNUNuZXhwb3J0JTIwKiUyMGZyb20lMjAnZXh0ZXJuYWwnJTNCJTIyJTJDJTIyaXNFbnRyeSUyMiUzQWZhbHNlJTJDJTIybmFtZSUyMiUzQSUyMmRlcDEuanMlMjIlN0QlNUQlMkMlMjJvcHRpb25zJTIyJTNBJTdCJTIyb3V0cHV0JTIyJTNBJTdCJTIyZm9ybWF0JTIyJTNBJTIyZXMlMjIlMkMlMjJpbmxpbmVEeW5hbWljSW1wb3J0cyUyMiUzQXRydWUlN0QlN0QlN0Q=